File: clients/typescript/src/topic.ts
import { ulid } from 'ulid';
import { connectionManager } from './connection_manager';
import { Message, TopicMessage } from './dtos';
import { resourceManager } from './resource_manager';
import { clientSetup, postJson, AmenoException } from './shared';
import { ITopic } from './topic_manager';

export type TopicHandler<T> = (message: TopicMessage<T>) => void;

export class Topic<T> implements ITopic {
    public readonly name: string;
    private disposed = false;
    private readonly handlers: Set<TopicHandler<T>> = new Set();

    constructor(name: string) {
        this.name = name;
    }

    public subscribe(handler: TopicHandler<T>): void {
        this.ensureNotDisposed();
        this.handlers.add(handler);
        connectionManager.topicManager.subscribeTopic(this);
    }

    public async publish(value: T, options?: { previous?: Message | null }): Promise<void> {
        this.ensureNotDisposed();

        const topicMessage: TopicMessage<T> = {
            id: ulid(),
            previous: options?.previous ?? null,
            type: 'TopicMessage',
            createdAt: new Date().toISOString(),
            topic: this.name,
            payload: value,
            appName: clientSetup.appName
        };

        const url = `${clientSetup.originUrl}/api/topic/publish`;
        await postJson(url, topicMessage);
    }

    public dispose(): void {
        this.ensureNotDisposed();
        this.disposed = true;
        this.handlers.clear();
        connectionManager.topicManager.unsubscribeTopic(this);
    }

    public dispatchMessage(message: TopicMessage<unknown>): void {
        if (this.disposed)
            return;

        const typedMessage = message as TopicMessage<T>;
        for (const handler of Array.from(this.handlers))
            handler(typedMessage);
    }

    private ensureNotDisposed(): void {
        if (this.disposed)
            throw new AmenoException(`O tópico '${this.name}' já foi descartado (disposed).`);
    }
}

export function topic<T>(name: string): Topic<T> {
    resourceManager.topics.add(name);
    return new Topic<T>(name);
}

File: clients/csharp/AmenoLink/Topic.cs
namespace AmenoLink;

public class Topic<T>(string name) : ITopic<T>, ITopicSubscriber
{
    public string Name { get; } = name;
    private bool disposed;
    private readonly HashSet<Func<TopicMessage<T>, Task>> handlers = [];
    private readonly Lock lockObject = new();

    public void Subscribe(Func<TopicMessage<T>, Task> handler)
    {
        EnsureNotDisposed();
        lock (lockObject)
            handlers.Add(handler);

        AmenoLinkClient.Connection.TopicManager.SubscribeTopic(this);
    }

    public async Task Publish(T value, Message? previous = null)
    {
        EnsureNotDisposed();

        var message = new TopicMessage<T>(
            Id: Ulid.NewUlid().ToString(),
            CreatedAt: DateTime.UtcNow.ToString("o"),
            Topic: Name,
            Type: "TopicMessage",
            Payload: value,
            Previous: previous,
            AppName: AmenoLinkClient.Settings.AppName
        );

        string url = $"{AmenoLinkClient.Settings.OriginUrl}/api/topic/publish";
        await AmenoLinkClient.PostJson<object?>(url, message);
    }

    public void Dispose()
    {
        EnsureNotDisposed();
        disposed = true;

        lock (lockObject)
            handlers.Clear();

        AmenoLinkClient.Connection.TopicManager.UnsubscribeTopic(this);
        GC.SuppressFinalize(this);
    }

    void ITopicSubscriber.DispatchMessage(TopicMessage<object> message)
    {
        if (disposed)
            return;

        T? typedPayload = default;
        if (message.Payload is not null)
        {
            if (message.Payload is T payload)
                typedPayload = payload;
            else
            {
                var serialized = System.Text.Json.JsonSerializer.Serialize(message.Payload, AmenoLinkClient.JsonOptions);
                typedPayload = System.Text.Json.JsonSerializer.Deserialize<T>(serialized, AmenoLinkClient.JsonOptions);
            }
        }

        var typedMessage = new TopicMessage<T>(
            Id: message.Id,
            CreatedAt: message.CreatedAt,
            Topic: message.Topic,
            Type: message.Type,
            Payload: typedPayload,
            Previous: message.Previous,
            AppName: message.AppName
        );

        List<Func<TopicMessage<T>, Task>> targets;
        lock (lockObject)
            targets = [.. handlers];

        foreach (var handler in targets)
            _ = handler(typedMessage);
    }

    private void EnsureNotDisposed()
    {
        if (disposed)
            throw new AmenoException($"O tópico '{Name}' já foi descartado (disposed).");
    }
}
